qsirecon/interfaces/gradients.py

```diff
@@ -18,33 +18,201 @@
+    shell_distances = pairwise_distances(select_shell_bs)
+    if np.any(shell_distances[np.triu_indices_from(shell_distances, k=1)] < max_distance):
+        raise Exception(
+            "Shells and bval_distance_cutoff have overlap. Choose a lower "
+            "bval_distance_cutoff or more separated shells."
+        )
+
+    selected_indices = np.flatnonzero(bval_df["assigned_shell"].isin(select_shell_bs))
+
+    return selected_indices
+
+
+def _parse_shell_selection(requested_bvals, bval_df, max_distance):
+    """Turn bval requests into numbers. Add a 0 if none were originally included."""
+    numeric_bvals = []
+    for requested_bval in requested_bvals:
+        if requested_bval == "highest":
+            highest_shell = bval_df["assigned_shell"].max()
+            LOGGER.info(f"Selecting b={highest_shell} as the highest shell")
+            numeric_bvals.append(highest_shell)
+        elif requested_bval == "lowest":
+            lowest_shell = bval_df["assigned_shell"][
+                bval_df["assigned_shell"] > max_distance
+            ].min()
+            LOGGER.info(f"Selecting b={lowest_shell} as the lowest b>0 shell")
+            numeric_bvals.append(lowest_shell)
+        else:
+            # Find the closest detected shell
+            ok_shells = bval_df["assigned_shell"][
+                np.logical_and(
+                    bval_df["assigned_shell"] >= requested_bval - max_distance,
+                    bval_df["assigned_shell"] <= requested_bval + max_distance,
+                )
+            ]
+            if len(ok_shells.unique()) > 1:
+                raise Exception(
+                    f"Unable to unambiguously select b={requested_bval}."
+                    f"instead select from {bval_df.assigned_shells.unique().tolist()}"
+                )
+
+            numeric_bvals.append(ok_shells.iloc[0])
+
+    # Make sure there is a 0 in the list
+    if len(numeric_bvals) < 2:
+        if numeric_bvals[0] < max_distance:
+            LOGGER.critical("No effectively b>0 shells are selected.")
+        else:
+            LOGGER.warning("Adding a b=0 shell to the selection.")
+            numeric_bvals.append(0)
+
+    return sorted(numeric_bvals)
+
+
+def _find_shells(bvals, max_distance):
+
+    import pandas as pd
+    from sklearn.cluster import AgglomerativeClustering
+
+    agg_cluster = AgglomerativeClustering(
+        n_clusters=None,
+        distance_threshold=2 * max_distance,
+        linkage="complete",
+    ).fit(bvals.reshape(-1, 1))
+
+    bval_df = pd.DataFrame({"bvalue": bvals, "assignment": agg_cluster.labels_})
+    shell_df = bval_df.groupby("assignment", as_index=False).agg({"bvalue": "median"})
+    bval_df["assigned_shell"] = bval_df["assignment"].replace(
+        shell_df["assignment"].tolist(), shell_df["bvalue"].tolist()
+    )
+    bval_df["shell_num"] = bval_df["assigned_shell"].rank(method="dense")
+    bval_df.drop("assigned_shell", inplace=True)
```

Comment:
```suggestion
    bval_df.drop(columns=["assigned_shell"], inplace=True)
```
I had to make this change to get it working locally.